### PR TITLE
Port search commands (`/`) to newshell

### LINKED ISSUE #1586: Port `/` (search) to rzshell
- [x] `/R` - ROP gadgets
- [x] `/` - string search
- [x] `/x` - hex search
- [x] `/a` - assembly search - `/a` done with https://github.com/rizinorg/rizin/pull/4931. Other commands: https://github.com/rizinorg/rizin/issues/4955
- [x] `/c` - cryptographic search - https://github.com/rizinorg/rizin/pull/4954
- [x] `/e` - regexp search
- [x] `/E` - ESIL search - https://github.com/rizinorg/rizin/pull/4982
- [x] `/F` - file search - https://github.com/rizinorg/rizin/pull/4972
- [x] `/g` - graph search - https://github.com/rizinorg/rizin/pull/4999
- [x] `/h` - hash search - https://github.com/rizinorg/rizin/pull/4995
- [x] `/m` - magic search - https://github.com/rizinorg/rizin/pull/4992
- [x] `/o` - offsets stuff - https://github.com/rizinorg/rizin/pull/4978
- [x] `/s` - sections search - https://github.com/rizinorg/rizin/pull/4995
- [x] `/p` - pattern search - https://github.com/rizinorg/rizin/pull/4983
- [x] `/P` - search similar blocks - https://github.com/rizinorg/rizin/pull/4983
- [x] `/v` - value search - https://github.com/rizinorg/rizin/pull/4975
- [x] `/V` - value range search - https://github.com/rizinorg/rizin/pull/4975
- [x] `/w` - wide string search
- [x] `/z` - string search

Note also the modifiers - `!`, `b`, `f`

### PR DESCRIPTION
**Pull request by DMaroo**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This PR ports the search commands (`/`) to Rizin newshell.

**Test plan**

Once the porting is done, the build and CI should be green. Maybe I'll add tests for any new features that get added during the porting.

**Closing issues**

Closes #1586.
Tracked by #1342.

### CODE REVIEW

File: librz/util/str.c [hunk 1]
```diff
@@ -2203,6 +2203,29 @@ RZ_API void rz_str_filter(char *str, int len) {
 	}
 }
 
+/**
+ * \brief Exttract all printable characters in \p str, and return a string of maximum length \p len
```
Comment (XVilka) [inline librz/util/str.c]: A typo

File: librz/util/str.c [hunk 2]
```diff
@@ -2203,6 +2203,29 @@ RZ_API void rz_str_filter(char *str, int len) {
...
+ * 
+ * \param str String to extract printable characters from
+ * \param len The maximum length of the resulting string
+ * \return char* Extracted printable string
+ */
+RZ_API char *rz_str_extract_printable(const char *str, int len) {
```
Comment (XVilka) [inline librz/util/str.c]: Add ownership mark for the return, `RZ_NONULL` to `str`

File: librz/util/str.c [hunk 3]
```diff
@@ -2203,6 +2203,29 @@ RZ_API void rz_str_filter(char *str, int len) {
...
+	rz_return_val_if_fail(str && len >= 0, NULL);
+	char *r, *res = malloc(len + 1);
+	int i;
```
Comment (XVilka) [inline librz/util/str.c]: Move `int i;` after error check

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/util/str.c [hunk 4]
```diff
@@ -2201,6 +2201,29 @@ RZ_API void rz_str_filter(char *str, int len) {
...
+ * \return char* Extracted printable string
+ */
+RZ_API RZ_OWN char *rz_str_extract_printable(RZ_NONNULL const char *str, int len) {
+	rz_return_val_if_fail(str && len >= 0, NULL);
+	char *r, *res = malloc(len + 1);
+	if (!res) {
+		return NULL;
+	}
+	int i;
+	for (i = 0, r = res; i < len; str++, i++) {
```
Comment (ret2libc) [inline librz/util/str.c]: Done this way it seems that len is the maximum length of `str` and not of the resulting string.

File: librz/core/cmd_descs/cmd_cmp.yaml [hunk 5]
```diff
@@ -157,6 +157,11 @@ commands:
     args:
       - name: hexpair
         type: RZ_CMD_ARG_TYPE_STRING
+    details:
+      - name: Ignoring bytes
+        entries:
+          - text: "Ignore specific bytes while comparing by using '.'"
+            comment: "cx 1ee7..0f....d34d"
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_cmp.yaml]: I suggest to use the "regular" `Examples` section.
Also, the command `cx` should co in `text`, the argument `1ee7..0f....d34d` in `arg_str` and `"Ignore specific bytes while comparing by using '.'"` in `comment`.

File: librz/core/cmd/cmd_print.c [hunk 6]
```diff
@@ -844,6 +844,36 @@ static const ut32 colormap[256] = {
 	0xffffff,
 };
 
+static void incAlphaBuffer(ut8 *buf, int bufsz) {
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c:634]: what are these?

File: librz/core/cmd/cmd_search.c [hunk 7]
```diff
@@ -438,13 +62,13 @@ static int _cb_hit(RzSearchKeyword *kw, void *user, ut64 addr) {
 			char *buf = calloc(1, len + 32 + ctx * 2);
 			type = "string";
 			rz_io_read_at(core->io, addr - prectx, (ut8 *)buf, len + (ctx * 2));
-			pre = getstring(buf, prectx);
-			pos = getstring(buf + prectx + len, ctx);
+			pre = rz_str_extract_printable(buf, prectx);
```
Comment (ret2libc) [inline librz/core/cmd/cmd_search.c:65]: add a unit test for the new rz_str_extract_printable API.

File: librz/core/cmd/cmd_search.c [hunk 8]
```diff
@@ -529,3338 +153,363 @@ static int _cb_hit(RzSearchKeyword *kw, void *user, ut64 addr) {
...
 
 static inline void print_search_progress(ut64 at, ut64 to, int n, struct search_parameters *param) {
-	if ((++c % 64) || (param->outmode == RZ_MODE_JSON)) {
+	if ((++progress % 64) || (param->outmode == RZ_MODE_JSON)) {
 		return;
 	}
 	if (rz_cons_singleton()->columns < 50) {
-		eprintf("\r[  ]  0x%08" PFMT64x "  hits = %d   \r%s",
-			at, n, (c % 2) ? "[ #]" : "[# ]");
+		rz_cons_printf("\r[  ]  0x%08" PFMT64x "  hits = %d   \r%s",
```
Comment (ret2libc) [inline librz/core/cmd/cmd_search.c]: i'd use RZ_LOG_INFO because this is not the real output of the command, it's just some update on the status.

**Review by ret2libc — COMMENTED** [review]
I think the RzSearchParameters split make sense, but I'd have a new `_new` API that accepts a `RzSearchParameters` structure to initialize the search. I think starting writing unit tests will definitely help to find out what is the best way to deal with this.

I'd see something like the following:
```C
RzSearchParams params = {
.mode = RZ_SEARCH_STRING,
.string_min = 3,
.string_max = 10,
};
RzSearch *s = rz_search_new(&params);
```

note that some of the parameters may be meaningful only for a specific "search mode". In such cases it may make sense to have RzSearchParameters composed of a union for the data, selected by the `search_mode` field.

File: librz/include/rz_search.h [hunk 9]
```diff
@@ -51,39 +51,62 @@ typedef struct rz_search_hit_t {
...
-	int align;
-	int (*update)(struct rz_search_t *s, ut64 from, const ut8 *buf, int len);
-	RzList *kws; // TODO: Use rz_search_kw_new ()
-	RzIOBind iob;
-	char bckwrds;
+	RzList /*<RzSearchHit>*/ *hits;
+	RzSearchParams *params;
 } RzSearch;
 
+typedef struct rz_core_t RzCore;
```
Comment (ret2libc) [inline librz/include/rz_search.h]: This is not good here. There should be no reference to core

Comment (DMaroo) [reply librz/include/rz_search.h]: Yeah, it's no longer necessary. I'll remove it.

File: librz/core/cmd/cmd_print.c (hunk 6 above)
Comment (DMaroo) [reply librz/core/cmd/cmd_print.c:634]: These are some functions that I needed to keep such that the build passes, I will refactor them in the end.

**Comment by XVilka** [comment]
It needs rebase. If you finish some particular command or few of them - probably better to focus on that, so the porting is incremental. Otherwise, we have to postpone this for 0.6.0

**Comment by XVilka** [comment]
@DMaroo do you plan to continue working on this? Otherwise it probably better to close this one and start from scratch.

**Comment by DMaroo** [comment]
@XVilka I have been very low on time lately. So I don't expect any progress on this for a while. Closing it seems reasonable. I'll open a new PR whenever I get done with the porting.

**Comment by DMaroo** [comment]
Changes have become stale and there are too many conflicts. Closing this PR for now. ETA on this is undecided.

**Comment by ret2libc** [comment]
I'm going to work on this as discussed on mattermost. I will start with `/R` and then expand from there. I will just try to convert the commands for now and leave the RzSearch refactoring to later.

**Pull request closed without merging**